### ISSUE TITLE
VM doesn't recognize top-level function as const in all cases.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#27164]
Example:
```dart
void main() {
  print(const Echo().echo("test"));
}


[12:08:17]lrn@lrn:~/dart/co/sdk/dev (reapply-fix-remove-resource)
typedef String EchoType(String arg);

const EchoType cecho = simpleEcho;

class Echo {
  final EchoType echo;
  const Echo() : echo = simpleEcho;
  const Echo.verbose() : echo = verboseEcho;
}

String simpleEcho(String arg) => arg;

String verboseEcho(String arg) => "arg: $arg";

void main() {
//  print(cecho("c"));
  print(const Echo().echo("test"));
}
```
This program fails with the error:
```
'file:///.../echo.dart': error: line 21 pos 21: Invalid const object field: Closure: (String) => String from Function 'simpleEcho': static.

  print(const Echo().echo("test"));
                    ^
```
It fails to recognize that `simpleEcho` is a compile-time constant in initializer list of `Echo()`.

However, if the commented out line is enabled, the `simpleEcho` function is recognized as constant, even in the following line.

The `simpleEcho` function is a compile-time function in all cases.

**Comment by mhausner** [comment]
Fixed in c15bac4812532215d0e286ce128493121e003fde

**Comment by mhausner** [comment]
FWIW, this was another manifestation of the imperfect implementation of "potentially const values", issue #392. This fix makes it a bit less imperfect :)
